File: backend/node_version/src/services/tokenService.ts
// server/src/services/tokenService.ts
import axios from 'axios';
import { OAuth2Client } from 'google-auth-library';


// Verify Google OAuth Access Token by calling Google People API
export const verifyGoogleToken = async (token: string) => {
  try {
    // Use Google People API to get user info
    const response = await axios.get('https://www.googleapis.com/oauth2/v3/userinfo', {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    if (response.status !== 200) {
      throw new Error('Invalid Google OAuth token.');
    }

    const { email, name } = response.data; // Extract user info from the response
    return { email, name }; // Return the user info
  } catch (error) {
    throw new Error('Failed to verify Google OAuth token.');
  }
};

// Verify Microsoft Token
export const verifyMicrosoftToken = async (token: string) => {
  const response = await axios.get('https://graph.microsoft.com/v1.0/me', {
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  if (response.status !== 200) {
    throw new Error('Invalid Microsoft token.');
  }

  const { mail, displayName } = response.data; // Extract user information from the response
  return { email: mail, name: displayName }; // Return extracted information
};

File: backend/scala-seed-project/src/services/TokenService.scala
// src/services/TokenService.scala

import cats.effect.IO
import sttp.client._
import sttp.client.asynchttpclient.cats._
import io.circe.generic.auto._
import io.circe.parser._

case class GoogleUserInfo(email: String, name: String)

object TokenService {
  
  private implicit val backend = AsyncHttpClientCatsBackend()

  // Verify Google OAuth Access Token by calling Google People API
  def verifyGoogleToken(token: String): IO[GoogleUserInfo] = {
    val request = basicRequest
      .header("Authorization", s"Bearer $token")
      .get(uri"https://www.googleapis.com/oauth2/v3/userinfo")

    for {
      response <- request.send()
      userInfo <- response.body match {
        case Right(body) =>
          decode[GoogleUserInfo](body) match {
            case Right(info) => IO.pure(info)
            case Left(_) => IO.raiseError(new Exception("Failed to decode Google user info."))
          }
        case Left(_) => IO.raiseError(new Exception("Invalid Google OAuth token."))
      }
    } yield userInfo
  }

  // Verify Microsoft Token
  def verifyMicrosoftToken(token: String): IO[GoogleUserInfo] = {
    val request = basicRequest
      .header("Authorization", s"Bearer $token")
      .get(uri"https://graph.microsoft.com/v1.0/me")

    for {
      response <- request.send()
      userInfo <- response.body match {
        case Right(body) =>
          decode[GoogleUserInfo](body) match {
            case Right(info) => IO.pure(info)
            case Left(_) => IO.raiseError(new Exception("Failed to decode Microsoft user info."))
          }
        case Left(_) => IO.raiseError(new Exception("Invalid Microsoft token."))
      }
    } yield userInfo
  }
}
